sanity-io/structured-content-2022
# Set up Live Preview functionality

# Set up Live Preview functionality

## Intent

- Implement support for [Live Preview](https://www.sanity.io/guides/nextjs-live-preview) 
- Fix [this Shortcut Story](https://app.shortcut.com/sanity-io/story/14246/set-up-preview-in-studio)

## Description

This PR implements Live Preview support for [Route](https://structured-content-2022-studio-1plwyyvga.sanity.build/desk/route) entities in Studio by following the linked guide. Assuming the general setup in this PR is OK, a follow-up PR can expand `resolveProductionUrl` to also work for `page`-type documents, but this would require a bit of GROQ cleverness which is out of scope for this PR.

Note the changed placement of the `...,` operator in the `QUERY` in _web/pages/[[...slug]].tsx_; I spent a good while trying to figure out why the client-side query errored out before realizing that the server- and client-side Sanity clients interpret the results of such a positioned `...` operator differently. Moving this operator to the top fixed the issue. (cc @kmelve, maybe fixing this upstream would save others some potential pain.)

## Testing this PR

1. Add a _.env.development_ file to the root of your repo (_.env_ files are not picked up!)
2. Enter `SANITY_STUDIO_PREVIEW_SECRET='<VALUE>'` into the file, replacing `<VALUE>` with the value of the [environment variable set up in Vercel](https://vercel.com/sanity-io/structured-content-2022-studio/settings/environment-variables).
3. Start the _/studio_ and _/web_ projects separately (`cd /[web/studio] && yarn && yarn dev`)
4. Navigate to e.g. [/shared-sections-test](http://localhost:3000/shared-sections-test) and verify that you get additional draft content (e.g., I added a "This is a test page" Hero heading to the [Studio document](https://structured-content-2022-studio-1plwyyvga.sanity.build/desk/page;345118ad-3b89-49c9-9ef1-c98f1db78cce)) loaded a couple seconds after the initial load
5. Open Studio locally and navigate to e.g. the [/shared-section-test](http://localhost:3333/desk/route;36c8bda6-e3a3-4671-a260-cdaf3dcd2e43) document
6. Click the top-right meatballs menu and verify that a "Open preview" option is present
7. Click said option, and verify that you're taken to http://localhost:3000/shared-sections-test, which again should load draft content post-load
8. Change some content in the document and verify that the content of the page you opened in your browser changes too


("Open preview" does not redirect to correct URLs for Preview Deploys, instead redirecting to the staging API (where this functionality is not yet available. We also need to remember to update this URL post-launch to point to production.)
